drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1328 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Driver for Nordic Semiconductor nRF I2S
+ */
+
+
+#include <stdlib.h>
+#include <i2s.h>
+#include <nrfx_i2s.h>
+#include "i2s_nrfx.h"
+#include <logging/log.h>
+
+#define LOG_LEVEL _LOG_LEVEL_DBG
+LOG_MODULE_REGISTER(i2s_nrfx, CONFIG_I2S_LOG_LEVEL);
+
+#define LOG_ERROR(msg) LOG_ERR("\r\n[%s:%u]: %s\r\n", __func__, __LINE__, msg)
+#define RING_BUF_INC(idx, limit) {idx = (++idx < limit) ? idx : 0; }
+
+#define DEV_CFG(dev) \
+	(const struct i2s_nrfx_config *const)((dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2s_nrfx_data *const)(dev)->driver_data)
+
+struct i2s_nrfx_interface;
+
+enum i2s_if_state {
+	I2S_IF_NOT_READY = 0,
+	I2S_IF_READY,
+	I2S_IF_RESTARTING,
+	I2S_IF_RUNNING,
+	I2S_IF_STOPPING,
+	I2S_IF_NEEDS_RESTART,
+	I2S_IF_ERROR
+};
+
+struct i2s_nrfx_config {
+	u8_t sck_pin;
+	u8_t lrck_pin;
+	u8_t mck_pin;
+	u8_t sdout_pin;
+	u8_t sdin_pin;
+	void (*instance_init)(struct device *dev);
+};
+
+struct queue_item {
+	void *data;
+	u32_t size;
```

Comment:
Perhaps I'm missing something, but ...

- for the TX queue:
  - when an item is queued, this field is initialized with the value passed through the `size` parameter to the `i2s_write` function (which is implemented as `i2s_nrfx_write`)
  - when an item is fetched from the queue, the value from this field is ignored (it is fetched to a variable called `mem_block_size`, but this variable is not used later on, see [`channel_tx_start`](https://github.com/zephyrproject-rtos/zephyr/pull/19436/files#diff-19d236d1c1e45c95ec25a37957e0cd66R827) as an example, other fetches look the same)

- for the RX queue:
  - this field is always initialized to the value equal to `i2s->size` (see [`channel_rx_callback`](https://github.com/zephyrproject-rtos/zephyr/pull/19436/files#diff-19d236d1c1e45c95ec25a37957e0cd66R1144)), which in turn is initialized to the `block_size` value for the first configured channel, see [`cfg_periph_config`](https://github.com/zephyrproject-rtos/zephyr/pull/19436/files#diff-19d236d1c1e45c95ec25a37957e0cd66R509) (BTW is it possible to change the block size once it is configured? It does not seem so in this implementation, but the API does not say that it is not supposed to be changed)

My question is: what's the point in storing this field for each queue item?

Another thing is that the current implementation silently ignores the value of the `size` parameter passed to the `i2s_write` function and just uses the configured memory block size, while the API states that this value [can be smaller that the size of the memory block](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/i2s.h#L459).